RZDataBinding/RZDBCoalesce.m

```diff
@@ -0,0 +1,317 @@
+    // if there are any arguments, the first and only argument must be a dictionary
+    if ( methodSig.methodReturnLength == 0 && (methodSig.numberOfArguments == 2 || methodSig.numberOfArguments == 3) ) {
+        notification = [[self alloc] init];
+        notification.target = invocation.target;
+        notification.action = invocation.selector;
+
+        if ( methodSig.numberOfArguments == 3 && strcmp([methodSig getArgumentTypeAtIndex:2], @encode(id)) == 0 ) {
+            __unsafe_unretained id arg1;
+            [invocation getArgument:&arg1 atIndex:2];
+
+            if ( [arg1 isKindOfClass:[NSDictionary class]] ) {
+                notification.changeDict = [arg1 mutableCopy];
+            }
+            else {
+                notification = nil;
+            }
+        }
+    }
+
+    return notification;
+}
+
+- (BOOL)isEqual:(id)object
+{
+    BOOL equal = [super isEqual:object];
+
+    if ( !equal && [object isKindOfClass:[RZDBNotification class]] ) {
+        RZDBNotification *other = (RZDBNotification *)object;
+
+        equal = (self.target == other.target) &&
+        (self.action == other.action) &&
+        (self.changeDict == other.changeDict ||
+         (self.changeDict[kRZDBChangeKeyObject] == other.changeDict[kRZDBChangeKeyObject] &&
+          [self.changeDict[kRZDBChangeKeyKeyPath] isEqualToString:other.changeDict[kRZDBChangeKeyKeyPath]]
+          )
+         );
+    }
+
+    return equal;
+}
+
+- (void)send
+{
+    if ( self.target != nil && self.action != NULL ) {
+        if ( self.changeDict != nil ) {
+            ((void(*)(id, SEL, NSDictionary *))objc_msgSend)(self.target, self.action, [self.changeDict copy]);
+        }
+        else {
+            ((void(*)(id, SEL))objc_msgSend)(self.target, self.action);
+        }
+    }
+}
+
+@end
+
+#pragma mark - NSObject+RZDBCoalesce implementation
+
+@implementation NSObject (RZDBCoalesce)
+
+- (id)rz_coalesceProxy
+{
+    id proxy = objc_getAssociatedObject(self, _cmd);
+
+    if ( proxy == nil ) {
+        proxy = [RZDBCoalesceProxy proxyForObject:self];
+        objc_setAssociatedObject(self, _cmd, proxy, OBJC_ASSOCIATION_RETAIN_NONATOMIC);
+    }
+
+    return proxy;
+}
+
+@end
+
+#pragma mark - RZDBCoalesceProxy implementation
+
+@implementation RZDBCoalesceProxy
+
++ (instancetype)proxyForObject:(id)object
+{
+    RZDBCoalesceProxy *proxy = [self alloc];
+    proxy.representedObject = object;
+
+    return proxy;
+}
+
+- (NSMethodSignature *)methodSignatureForSelector:(SEL)sel
+{
+    return [self.representedObject methodSignatureForSelector:sel];
+}
+
+- (void)forwardInvocation:(NSInvocation *)invocation
+{
+    invocation.target = self.representedObject;
+
+    RZDBCoalesce *currentCoalesce = [RZDBCoalesce currentCoalesceBeginIfNeeded:YES];
+
+    if ( currentCoalesce != nil ) {
+        // notificaton will be non-nil for a valid RZDataBinding callback
+        RZDBNotification *notification = [RZDBNotification notificationWithInvocation:invocation];
```

Comment:
It was because of auto-coalesce, even if it's empty now it may not be empty by the end of the runloop. But yeah this oddity will go away when I take out auto-coalesce